File: pages/20260824170700_holding_topology_passing/analysis/julia/passing_notation.jl
function skip_space(text, index)
    while index <= lastindex(text) && isspace(text[index])
        index = nextind(text, index)
    end
    return index
end

function throw_height(character)
    '0' <= character <= '9' && return character - '0'
    'a' <= character <= 'z' && return 10 + (character - 'a')
    throw(ArgumentError("unsupported passing throw: $character"))
end

function parse_passing_throw(text, index)
    index = skip_space(text, index)
    index > lastindex(text) && throw(ArgumentError("unexpected end of passing throw"))
    height = throw_height(text[index])
    index += 1
    crossing = isodd(height)
    if index <= lastindex(text) && (text[index] == 'x' || text[index] == 'X')
        crossing = !crossing
        index += 1
    end
    pass = false
    pass_target = nothing
    if index <= lastindex(text) && (text[index] == 'p' || text[index] == 'P')
        pass = true
        index += 1
        if index <= lastindex(text) && '1' <= text[index] <= '9'
            digits = Char[]
            while index <= lastindex(text) && '0' <= text[index] <= '9'
                push!(digits, text[index])
                index += 1
            end
            pass_target = parse(Int, String(digits))
        end
    end
    return ThrowToken(height, crossing, pass, pass_target), index
end

function parse_multiplex(text, index)
    throws = ThrowToken[]
    index += 1
    while index <= lastindex(text) && text[index] != ']'
        index = skip_space(text, index)
        if text[index] == '/'
            index += 1
            continue
        end
        text[index] == ']' && break
        token, index = parse_passing_throw(text, index)
        push!(throws, token)
    end
    index <= lastindex(text) && text[index] == ']' || throw(ArgumentError("unclosed passing multiplex"))
    isempty(throws) && throw(ArgumentError("empty passing multiplex"))
    return throws, index + 1
end

function parse_section(text)
    throws = Vector{ThrowToken}[]
    starting_hand = nothing
    index = firstindex(text)
    while index <= lastindex(text)
        index = skip_space(text, index)
        index > lastindex(text) && break
        mark = text[index]
        if (mark == 'R' || mark == 'L') && isempty(throws) && starting_hand === nothing
            starting_hand = mark == 'R' ? 1 : 0
            index += 1
            continue
        end
        if mark == '['
            multiplex, index = parse_multiplex(text, index)
            push!(throws, multiplex)
            continue
        end
        token, index = parse_passing_throw(text, index)
        push!(throws, [token])
    end
    return (starting_hand=starting_hand, throws=throws)
end

function parse_angle_block(text, index)
    text[index] == '<' || throw(ArgumentError("passing notation expected <"))
    close = findnext('>', text, index)
    close === nothing && throw(ArgumentError("unclosed passing block"))
    inner = text[(index + 1):(close - 1)]
    parts = NamedTuple[]
    depth = 0
    start = 1
    for cursor in 1:(length(inner) + 1)
        character = cursor <= length(inner) ? inner[cursor] : '\0'
        depth += Int(character == '[' || character == '(')
        depth -= Int(character == ']' || character == ')')
        if cursor == length(inner) + 1 || (character == '|' && depth == 0)
            push!(parts, parse_section(inner[start:(cursor - 1)]))
            start = cursor + 1
        end
    end
    length(parts) < 2 && throw(ArgumentError("passing block needs at least two bodies"))
    return parts, close + 1
end

function merge_blocks(blocks)
    body_count = length(blocks[1])
    all(length(block) == body_count for block in blocks) ||
        throw(ArgumentError("passing blocks must keep the same body count"))
    starting_hands = fill(1, body_count)
    throws = [Vector{ThrowToken}[] for _ in 1:body_count]
    for block in blocks
        for body in 1:body_count
            section = block[body]
            if section.starting_hand !== nothing
                starting_hands[body] = section.starting_hand
            end
            append!(throws[body], section.throws)
        end
    end
    beat_count = length(throws[1])
    all(length(sequence) == beat_count for sequence in throws) ||
        throw(ArgumentError("each body must contribute the same number of beats"))
    beat_count == 0 && throw(ArgumentError("passing pattern has no throws"))
    return PassingPattern(body_count, starting_hands, throws)
end

function parse_passing_siteswap(source)
    text = strip(String(source))
    blocks = Vector{Vector{NamedTuple}}()
    index = firstindex(text)
    while index <= lastindex(text)
        index = skip_space(text, index)
        index > lastindex(text) && break
        block, index = parse_angle_block(text, index)
        push!(blocks, block)
    end
    isempty(blocks) && throw(ArgumentError("empty passing siteswap"))
    return merge_blocks(blocks)
end

function passing_object_count(pattern::PassingPattern)
    height_sum = sum(sum(sum(token.height for token in multiplex) for multiplex in sequence)
                     for sequence in pattern.throws)
    return height_sum ÷ length(pattern.throws[1])
end

File: pages/20260824170700_holding_topology_passing/analysis/ruby/passing_notation.rb
module PassingAnalysis
  ThrowToken = Struct.new(:height, :crossing, :pass, :pass_target, keyword_init: true)
  Pattern = Struct.new(:body_count, :starting_hands, :throws, keyword_init: true)

  module Notation
    module_function

    def parse(source)
      text = source.to_s.strip
      raise ArgumentError, "empty passing siteswap" if text.empty?

      blocks = []
      index = 0
      while index < text.length
        index = skip_space(text, index)
        break if index >= text.length

        block, index = parse_angle_block(text, index)
        blocks << block
      end
      raise ArgumentError, "empty passing siteswap" if blocks.empty?

      merge_blocks(blocks)
    end

    def object_count(pattern)
      height_sum = pattern.throws.sum do |sequence|
        sequence.sum { |multiplex| multiplex.sum(&:height) }
      end
      height_sum / pattern.throws.first.length
    end

    def skip_space(text, index)
      index += 1 while index < text.length && text[index] =~ /\s/
      index
    end

    def parse_throw(text, index)
      index = skip_space(text, index)
      raise ArgumentError, "unexpected end of passing throw" if index >= text.length

      height = throw_height(text[index])
      index += 1
      crossing = height.odd?
      if text[index] == "x" || text[index] == "X"
        crossing = !crossing
        index += 1
      end
      pass = false
      pass_target = nil
      if text[index] == "p" || text[index] == "P"
        pass = true
        index += 1
        if text[index] =~ /[1-9]/
          digits = +""
          while text[index] =~ /[0-9]/
            digits << text[index]
            index += 1
          end
          pass_target = Integer(digits)
        end
      end
      [ThrowToken.new(height: height, crossing: crossing, pass: pass, pass_target: pass_target), index]
    end

    def throw_height(character)
      return Integer(character) if character =~ /[0-9]/
      return 10 + character.ord - 97 if character =~ /[a-z]/

      raise ArgumentError, "unsupported passing throw: #{character}"
    end

    def parse_multiplex(text, index)
      throws = []
      index += 1
      while index < text.length && text[index] != "]"
        index = skip_space(text, index)
        if text[index] == "/"
          index += 1
          next
        end
        break if text[index] == "]"

        token, index = parse_throw(text, index)
        throws << token
      end
      raise ArgumentError, "unclosed passing multiplex" unless text[index] == "]"
      raise ArgumentError, "empty passing multiplex" if throws.empty?

      [throws, index + 1]
    end

    def parse_section(text)
      throws = []
      starting_hand = nil
      index = 0
      while index < text.length
        index = skip_space(text, index)
        break if index >= text.length

        mark = text[index]
        if %w[R L].include?(mark) && throws.empty? && starting_hand.nil?
          starting_hand = mark == "R" ? 1 : 0
          index += 1
          next
        end
        if mark == "["
          multiplex, index = parse_multiplex(text, index)
          throws << multiplex
          next
        end
        token, index = parse_throw(text, index)
        throws << [token]
      end
      { starting_hand: starting_hand, throws: throws }
    end

    def parse_angle_block(text, index)
      raise ArgumentError, "passing notation expected <" unless text[index] == "<"

      close = text.index(">", index)
      raise ArgumentError, "unclosed passing block" if close.nil?

      inner = text[(index + 1)...close]
      parts = []
      depth = 0
      start = 0
      (0..inner.length).each do |cursor|
        character = inner[cursor]
        depth += 1 if character == "[" || character == "("
        depth -= 1 if character == "]" || character == ")"
        next unless cursor == inner.length || (character == "|" && depth.zero?)

        parts << parse_section(inner[start...cursor])
        start = cursor + 1
      end
      raise ArgumentError, "passing block needs at least two bodies" if parts.length < 2

      [parts, close + 1]
    end

    def merge_blocks(blocks)
      body_count = blocks.first.length
      raise ArgumentError, "passing blocks must keep the same body count" unless
        blocks.all? { |block| block.length == body_count }

      starting_hands = Array.new(body_count, 1)
      throws = Array.new(body_count) { [] }
      blocks.each do |block|
        block.each_with_index do |section, body|
          starting_hands[body] = section.fetch(:starting_hand) unless section.fetch(:starting_hand).nil?
          throws[body].concat(section.fetch(:throws))
        end
      end
      beat_count = throws.first.length
      raise ArgumentError, "each body must contribute the same number of beats" unless
        throws.all? { |sequence| sequence.length == beat_count }
      raise ArgumentError, "passing pattern has no throws" if beat_count.zero?

      Pattern.new(body_count: body_count, starting_hands: starting_hands, throws: throws)
    end
  end
end
